src/cli.cr

```diff
@@ -2,7 +2,80 @@ require "dppm/cli"
 require "./dppm_rest_api"
 
 module DppmRestApi::CLI
-  def self.run_server(config, host, port, data_dir, **args)
+  extend self
+
+  # A helper method for the select_users method -- splits a comma-separated
+  # list of numbers in a string into an `Array` of `Int32` values.
+  private def split_groups_arg(argument text : String) : Array(Int32)
+    text.split(',').map do |g|
+      if g_int = g.to_i?
+        g_int
+      else
+        abort "\
+          group ID #{g} wasn't a number, please use numeric group ids \
+          separated by commas (,)."
+      end
+    end
+  end
+
+  # All variables are received by a method as a `String | Nil`. This macro
+  # converts an argument, by name, to its `String` value, or raises an
+  # error instructing the user to include the variable value.
+  private macro required(arg)
+    mn_{{arg.id}} || abort "{{ arg.id.gsub /_/, "-" }} argument is required."
+  end
+
+  # select the users from the current configuration according to the name,
+  # groups the user is a member of, and/or the user's API key.
+  private def select_users(match_name, match_groups, api_key, from selected_users)
+    # Filter by name
+    if name = match_name
+      if name.starts_with?('/') && name.ends_with?('/')
+        # match by regex
+        name_regex = Regex.new name.lchop('/').rchop('/')
+        selected_users = selected_users.select do |user|
+          name_regex =~ user.name
+        end
+      else
+        # find users by exact match
+        selected_users = selected_users.select do |user|
+          user.name == name
+        end
+      end
+    end
+    # Filter by group membership
+    if groups = match_groups
+      group_matches = Set(Set(Int32)).new
+      groups.split(':').each do |part|
+        group_matches << split_groups_arg(part).to_set
+      end
+      selected_users = selected_users.select do |user|
+        group_matches.any? do |group_set|
+          group_set.all? do |group|
+            user.group_ids.includes? group
+          end
+        end
+      end
+    end
+    # filter by API key if specified.
+    if key = api_key
+      selected_users = selected_users.select { |user| user.api_key_hash == key }
+      # freak out if multiple users were found.
+      if selected_users.size > 1
+        abort String.build { |msg|
```

Comment:
Use `do ... end`

Reply:
I used curly-braces in this case because it has different precedence than `do..end` I will save as an intermediate variable to avoid confusion